pkg/scm/provider/github.go

```diff
@@ -0,0 +1,185 @@
+/*
+Copyright 2017 caicloud authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package provider
+
+import (
+	"fmt"
+
+	"github.com/caicloud/cyclone/cloud"
+	"github.com/caicloud/cyclone/pkg/api"
+	"github.com/caicloud/cyclone/pkg/osutil"
+	httperror "github.com/caicloud/cyclone/pkg/util/http/errors"
+	"github.com/caicloud/cyclone/store"
+	"github.com/google/go-github/github"
+	"golang.org/x/oauth2"
+	mgo "gopkg.in/mgo.v2"
+)
+
+// GitHubManager represents the manager for github.
+type GitHubManager struct {
+	DataStore *store.DataStore
+}
+
+// Authcallback is the callback handler.
+func (g *GitHubManager) Authcallback(code, state string) (string, error) {
+	if code == "" || state == "" {
+		return "", fmt.Errorf("code: %s or state: %s is nil", code, state)
+	}
+
+	// Caicloud web address,eg caicloud.io
+	uiPath := osutil.GetStringEnv(cloud.ConsoleWebEndpoint, "http://localhost:8000")
+	redirectURL := fmt.Sprintf("%s/cyclone/add?type=github&code=%s&state=%s", uiPath, code, state)
+
+	if err := g.setToken(code, state); err != nil {
+		return "", err
+	}
+
+	return redirectURL, nil
+}
+
+// setToken sets the token.
+func (g *GitHubManager) setToken(code, state string) error {
+
```

Comment:
Remove all empty lines before the code of functions, same errors in other places.

Reply:
ok